JSON api description: Support for fixed-size arrays and tuples

ascjones opened the pull request:
Fixes #40.

- Splits the top level `TypeDescription` enum into `Primitive`, `Array`, and `Tuple` variants.
- Adds some tests

Robbepop commented on lang/src/api.rs:
```diff
@@ -69,34 +69,72 @@ pub enum TypeDescription {
     /// The SRML balance type.
     Balance,
+    /// The tuple type
+    Tuple {
```
I just got the idea that we can even improve this further by doing something similar to how `syn` handles all its complex `enum` types: https://docs.rs/syn/0.15.29/syn/enum.Type.html

So we could define a `TupleTypeDescription` and `ArrayTypeDescription` where all the tuple- and array type description specific implementations are going. They would both implement their own `TryFrom` for `syn::TypeTuple` and `syn::TypeArray` respectively and we could have a much simpler `TryFrom` impl for `TypeDescription`.

Robbepop commented on lang/src/api.rs:
```diff
@@ -69,34 +69,72 @@ pub enum TypeDescription {
+                let elems = tuple
+                    .elems
+                    .iter()
+                    .map(primitive)
```
Interesting approach with the primitives map. However, as far as I can see this restricts us from having recursive types like `(u32, (bool, u32))` or `[[u32; 5]; 5]`. I think this is a simple fix by again moving the primitives into the root match and together with my suggestion above just simply call `TypeDescription::try_from` recursively in the `TupleTypeDescription` and `ArrayTypeDescription` respectively.

ascjones commented:
Implemented your suggestions and added some tests. The json ones are failing, I will need to customise the serde serialization for those variants to match the format specified in the issue.

ascjones commented:
To achieve the specified json representation I have:
- Extracted primitive types into their own enum (so they remain tagged)
- Annotated the top level `TypeDescription` enum so that it is `untagged`
- Made `ArrayDescription` a single case enum to allow renames

Note that in the last case the resulting json will be a json object e.g.
```
{
    "[T;n]": {
        "T": "u32",
        "n": "5"
    }
}
```
Assume this is correct because it will be the value of a type in the higher order structure.

Robbepop commented:
lgtm!

Now only the tuple-array mixed tests are missing for a merge. :)
